host/volume/api/http.go

```diff
@@ -0,0 +1,114 @@
+package volumeapi
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+
+	"github.com/flynn/flynn/Godeps/_workspace/src/github.com/julienschmidt/httprouter"
+	"github.com/flynn/flynn/host/volume"
+	"github.com/flynn/flynn/host/volume/zfs"
+	"github.com/flynn/flynn/pkg/httphelper"
+	"github.com/flynn/flynn/pkg/random"
+)
+
+type HttpAPI struct {
+	vman *volume.Manager
+}
+
+func NewHttpAPI(vman *volume.Manager) *HttpAPI {
+	return &HttpAPI{vman: vman}
+}
+
+func RegisterRoutes(api *HttpAPI, r *httprouter.Router) {
+	r.POST("/volume/provider", api.CreateProvider)
+	r.POST("/volume/provider/:provider_id/newVolume", api.Create)
+	r.PUT("/volume/instances/:id/snapshot", api.Snapshot)
+	r.GET("/volume/instances/:id/inspect", api.Inspect)
+}
+
+func (api *HttpAPI) CreateProvider(w http.ResponseWriter, r *http.Request, ps httprouter.Params) {
+	data, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		httphelper.Error(w, err)
+		return
+	}
+
+	var provider volume.Provider
+	pspec := &volume.ProviderSpec{}
+	if err = json.Unmarshal(data, &pspec); err != nil {
+		httphelper.Error(w, err)
+		return
+	}
+	switch pspec.Kind {
+	case "zfs":
+		config := &zfs.ProviderConfig{}
+		if err := json.Unmarshal(pspec.Config, config); err != nil {
+			httphelper.JSON(w, 400, errors.New("host: invalid config for zfs volume provider"))
+			return
+		}
+		var err error
+		if provider, err = zfs.NewProvider(config); err != nil {
+			httphelper.JSON(w, 500, err)
+			return
+		}
+	case "":
+		httphelper.JSON(w, 400, errors.New("host: volume provider kind must not be blank"))
+		return
+	default:
+		httphelper.JSON(w, 400, fmt.Errorf("host: volume provider kind '%s' is not known"))
```

Comment:
These errors should use the `JSONError` system we are currently working towards standardizing.